sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1278,47 +1277,33 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	if (!is_power_of_2(adr_link->mask))
 		return -EINVAL;
 
-	cpu_dai_id[index++] = ffs(adr_link->mask) - 1;
 	if (!adr_d->endpoints->aggregated || no_aggregation) {
+		cpu_dai_id[0] = ffs(adr_link->mask) - 1;
 		*cpu_dai_num = 1;
 		*codec_num = 1;
 		*group_id = 0;
 		return 0;
 	}
 
+	*codec_num = 0;
 	*group_id = adr_d->endpoints->group_id;
 
 	/* Count endpoints with the same group_id in the adr_link */
-	*codec_num = 0;
-	for (i = 0; i < adr_link->num_adr; i++) {
-		if (adr_link->adr_d[i].endpoints->aggregated &&
-		    adr_link->adr_d[i].endpoints->group_id == *group_id)
-			(*codec_num)++;
-	}
-
-	/* gather other link ID of slaves in the same group */
-	for (adr_next = adr_link + 1; adr_next && adr_next->num_adr;
-		adr_next++) {
-		const struct snd_soc_acpi_endpoint *endpoint;
-
-		endpoint = adr_next->adr_d->endpoints;
-		if (!endpoint->aggregated ||
-		    endpoint->group_id != *group_id)
-			continue;
-
+	for (; adr_link && adr_link->num_adr; adr_link++) {
 		/* make sure the link mask has a single bit set */
-		if (!is_power_of_2(adr_next->mask))
+		if (!is_power_of_2(adr_link->mask))
 			return -EINVAL;
 
 		if (index >= SDW_MAX_CPU_DAIS) {
 			dev_err(dev, " cpu_dai_id array overflows");
 			return -EINVAL;
 		}
 
-		cpu_dai_id[index++] = ffs(adr_next->mask) - 1;
-		for (i = 0; i < adr_next->num_adr; i++) {
-			if (adr_next->adr_d[i].endpoints->aggregated &&
-			    adr_next->adr_d[i].endpoints->group_id == *group_id)
+		cpu_dai_id[index++] = ffs(adr_link->mask) - 1;
+
+		for (i = 0; i < adr_link->num_adr; i++) {
+			if (adr_link->adr_d[i].endpoints->aggregated &&
+			    adr_link->adr_d[i].endpoints->group_id == *group_id)
```

Comment:
I'll let @bardliao  and @RanderWang review this less-than-simple piece of code :-)